code/modules/atmospherics/portable/pump.dm

```diff
@@ -109,81 +109,54 @@
 		return
 	return src.Attackhand(user)
 
-/obj/machinery/portable_atmospherics/pump/attack_hand(var/mob/user)
-
-	src.add_dialog(user)
-	var/holding_text
-
-	if(holding)
-		holding_text = {"<BR><B>Tank Pressure</B>: [MIXTURE_PRESSURE(holding.air_contents)] KPa<BR>
-<A href='?src=\ref[src];remove_tank=1'>Remove Tank</A><BR>
-"}
-	var/output_text = {"<TT><B>[name]</B><BR>
-Pressure: [MIXTURE_PRESSURE(air_contents)] KPa<BR>
-Port Status: [(connected_port)?("Connected"):("Disconnected")]
-[holding_text]
-<BR>
-Power Switch: <A href='?src=\ref[src];power=1'>[on?("On"):("Off")]</A><BR>
-Pump Direction: <A href='?src=\ref[src];direction=1'>[direction_out?("Out"):("In")]</A><BR>
-Target Pressure: <A href='?src=\ref[src];pressure_adj=-100'>-</A> <A href='?src=\ref[src];pressure_adj=-10'>-</A> <A href='?src=\ref[src];pressure_set=1'>[target_pressure]</A> <A href='?src=\ref[src];pressure_adj=10'>+</A> <A href='?src=\ref[src];pressure_adj=100'>+</A><BR>
-<HR>
-<A href='?action=mach_close&window=pump'>Close</A><BR>
-"}
-
-	user.Browse(output_text, "window=pump;size=600x300")
-	onclose(user, "pump")
-
-	return
-
-/obj/machinery/portable_atmospherics/pump/Topic(href, href_list)
-	if(..())
+/obj/machinery/portable_atmospherics/pump/ui_interact(mob/user, datum/tgui/ui)
+	ui = tgui_process.try_update_ui(user, src, ui)
+	if (!ui)
+		ui = new(user, src, "PortablePump", name)
+		ui.open()
+
+/obj/machinery/portable_atmospherics/pump/ui_data(mob/user)
+	. = list(
+		"pressure" = MIXTURE_PRESSURE(src.air_contents),
+		"on" = src.on,
+		"maxPressure" = src.maximum_pressure,
+		"connected" = src.connected_port ? TRUE : FALSE,
+		"targetPressure" = src.target_pressure,
+		"direction_out" = src.direction_out
+	)
+
+	.["holding"] = isnull(holding) ? null : list(
+		"name" = src.holding.name,
+		"pressure" = MIXTURE_PRESSURE(src.holding.air_contents),
+		"maxPressure" = PORTABLE_ATMOS_MAX_RELEASE_PRESSURE,
```

Comment:
maybe also static data, might be too difficult :S

Reply:
probably wont work